File: app/src/main/java/ca/mitenko/evn/model/search/DestSearch.java
package ca.mitenko.evn.model.search;

import android.location.Location;
import android.location.LocationManager;
import android.support.annotation.NonNull;

import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;

import org.immutables.value.Value;
import org.parceler.Parcel;
import org.parceler.ParcelFactory;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

import javax.annotation.Nullable;

import ca.mitenko.evn.model.Activity;
import ca.mitenko.evn.model.Destination;
import ca.mitenko.evn.model.ImmutableDestination;
import ca.mitenko.evn.util.LocationUtil;

/**
 * Created by mitenko on 2017-07-30.
 */

@Parcel(value = Parcel.Serialization.VALUE, implementations = ImmutableDestSearch.class)
@Value.Immutable
public class DestSearch {
    private static final LatLngBounds DEFAULT_BOUNDS = new
            LatLngBounds(
            new LatLng(49.15938572687397,-123.9760036021471),
            new LatLng(49.20606374369103,-123.91420517116785));

    /**
     * The search bounds. Default is roughly nanaimo
     * @return
     */
    @NonNull
    @Value.Default
    public LatLngBounds searchBounds() {
        return DEFAULT_BOUNDS;
    }

    /**
     * The map bounds. Default is roughly nanaimo
     * @return
     */
    @Nullable
    @Value.Default
    public LatLngBounds mapBounds() {
        return null;
    }

    /**
     * The map bounds. Default is roughly nanaimo
     * @return
     */
    @NonNull
    @Value.Lazy
    public LatLngBounds mapBoundsOrDefault() {
        if (mapBounds() != null) {
            return mapBounds();
        }
        return searchBounds();
    }

    /**
     * Returns true / false if the search
     * has a non-default set of map bounds
     * @return
     */
    @NonNull
    @Value.Lazy
    public boolean hasMapBounds() {
        return mapBounds() != null;
    }

    /**
     * Returns true / false if the map bounds are
     * outside of the search bounds
     * @return
     */
    @NonNull
    @Value.Lazy
    public boolean mapOutsideSearch() {
        if (!hasResults() || mapBounds() == null) {
            return false;
        }
        return !searchBounds().contains(mapBounds().northeast)
                || !searchBounds().contains(mapBounds().southwest);
    }

    /**
     * The filter to apply to the results
     * @return
     */
    @NonNull
    @Value.Default
    public Filter filter() {
        return ImmutableFilter.builder().build();
    }

    /**
     * The raw result set
     */
    @Nullable
    @Value.Default
    public ArrayList<Destination> results() {
        return null;
    }

    /**
     * The raw result set
     */
    @NonNull
    @Value.Lazy
    public boolean hasResults() {
        return results() != null;
    }

    /**
     * The filtered set of results
     */
    @NonNull
    @Value.Lazy
    @SuppressWarnings("ConstantConditions")
    public ArrayList<Destination> filteredResults(boolean filterByMapBounds) {
        if (!hasResults()) {
            return new ArrayList<>();
        }

        /**
         * Sort the list by nearest to farther if user location is known
         */
        ArrayList<Destination> orderedResults = results();
        if (filter().userLocation() != null) {
            Location userLocation = LocationUtil.fromLatLng(filter().userLocation());

            Collections.sort(orderedResults, new Comparator<Destination>() {
                public int compare(Destination d1, Destination d2) {
                    Location location1 = LocationUtil.fromDestination(d1);
                    Location location2 = LocationUtil.fromDestination(d2);
                    return Integer.signum(
                            (int)location1.distanceTo(userLocation) - (int)location2.distanceTo(userLocation));
                }
            });
        }

        /**
         * Filter by category and / or activity
         */
        if (filter().isEmpty()) {
            return orderedResults;
        }

        /**
         * Otherwise apply the filter
         */
        ArrayList<Destination> filteredResults = new ArrayList<>();
 /*       for(Destination destination : orderedResults) {
            Integer destinationCost = destination.getDetail().getCost();
            for (Activity activity : destination.getDetail().getActivities()) {
                *//**
                 * Filter by Activity / Cost / Category
                 *//*
                if ((filter().categories().isEmpty() || filter().categories().contains(activity.getCategory())) &&
                        (filter().activities().isEmpty() || filter().activities().contains(activity.getName())) &&
                        (filter().cost().isEmpty() || filter().cost().contains(destinationCost))) {
                    filteredResults.add(ImmutableDestination.builder()
                        .from(destination)
                        .displayIcon(activity.category())
                        .build());
                    break;
                }
            }
        }*/
        return filteredResults;
    }

    /**
     * Parcel Factory
     * @return
     */
    @ParcelFactory
    public static DestSearch build(Filter filter, LatLngBounds searchBounds, LatLngBounds mapBounds,
                                   ArrayList<Destination> results) {
        return ImmutableDestSearch.builder()
                .searchBounds(searchBounds)
                .mapBounds(mapBounds)
                .filter(filter)
                .results(results)
                .build();
    }
}

File: app/src/main/java/ca/mitenko/evn/model/search/DestSearch.kt
package ca.mitenko.evn.model.search

import android.os.Parcelable
import ca.mitenko.evn.model.Destination
import ca.mitenko.evn.util.LatLngUtil
import ca.mitenko.evn.util.LocationUtil
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import kotlinx.android.parcel.Parcelize
import org.immutables.value.Value
import timber.log.Timber
import java.util.*

/**
 * Created by mitenko on 2017-07-30.
 */

@Parcelize
data class DestSearch(
        val searchBounds: LatLngBounds = DEFAULT_BOUNDS,
        val mapBounds: LatLngBounds? = null,
        val filter: Filter = Filter(),
        val results: ArrayList<Destination>? = null

): Parcelable {
    /**
     * The map bounds. Default is roughly nanaimo
     * @return
     */
    fun mapBoundsOrDefault() = mapBounds ?: searchBounds

    /**
     * Returns true / false if the search
     * has a non-default set of map bounds
     * @return
     */
    fun hasMapBounds() = mapBounds != null

    /**
     * Returns true / false if the map bounds are
     * outside of the search bounds
     * @return
     */
    @Value.Lazy
    fun mapOutsideSearch(): Boolean {
        return if (!hasResults() || mapBounds == null) {
            false
        } else !searchBounds.contains(mapBounds.northeast) || !searchBounds.contains(mapBounds.southwest)
    }

    /**
     * The raw result set
     */
    fun hasResults() = results != null

    /**
     * The filtered set of results
     */
    fun filteredResults(filterByMapBounds: Boolean): ArrayList<Destination> {
        if (!hasResults()) {
            return ArrayList()
        }

        /**
         * Sort the list by nearest to farther if user location is known
         */
        var orderedResults = results ?: ArrayList()
        if (filter.userLocation != null) {
            val userLocation = LocationUtil.fromLatLng(filter.userLocation)

            orderedResults.sortWith(Comparator { d1, d2 ->
                val location1 = LocationUtil.fromDestination(d1)
                val location2 = LocationUtil.fromDestination(d2)
                Integer.signum(
                        location1.distanceTo(userLocation).toInt() - location2.distanceTo(userLocation).toInt())
            })
        }

        /**
         * Filter by Map Bounds
         */
        if (filterByMapBounds) {
            val bounds = mapBoundsOrDefault()
            orderedResults = orderedResults.filter {
                dest -> bounds.contains(LatLngUtil.fromDestination(dest))
            } as ArrayList<Destination>
        }

        /**
         * Filter by category and / or activity
         */
        if (filter.isEmpty) {
            return orderedResults
        }

        /**
         * Otherwise apply the filter
         */
        val filteredResults = ArrayList<Destination>()
        for (destination in orderedResults) {
            val destinationCost = destination.detail.cost
            for ((_, name, category) in destination.detail.activities) {
                /**
                 * Filter by Activity / Cost / Category
                 */
                if ((filter.categories.isEmpty() || filter.categories.contains(category)) &&
                        (filter.activities.isEmpty() || filter.activities.contains(name))) {
                    filteredResults.add(
                            destination.copy(displayIcon = category ?: "unknown"))
                    break
                }
            }
        }
        return filteredResults
    }

    companion object {
        val DEFAULT_BOUNDS = LatLngBounds(
                LatLng(49.15938572687397, -123.9760036021471),
                LatLng(49.20606374369103, -123.91420517116785))
    }
}
